### PR TITLE
1.0.0

### PR DESCRIPTION
**Pull request by JoshBBOXX**
# Changes
- Created AfricasTalkingProvider
- Implemented send functionality

# Tests
- Tests for send success and failure added

# Notes
- Repo should roughly emulate existing provider modules: [smsframework-clickatell], [smsframework-vianett]
- Provider works in conjunction with [smsframework]
- Functionality for premium sending, and receiving incoming messages to be added in future

[smsframework-clickatell]: https://github.com/kolypto/py-smsframework-clickatell
[smsframework-vianett]: https://github.com/kolypto/py-smsframework-vianett
[smsframework]: https://github.com/kolypto/py-smsframework

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,2 +1,3 @@
 africastalking==1.1.5
+phonenumbers==8.10.12
```
Comment (panoskar86) [inline requirements.txt]: I think in DOSH we have the light phonenumbers library. phonenumberslite==8.10.0
Check if we really need the full library

File: smsframework_africastalking/error.py [hunk 2]
```diff
@@ -0,0 +1,6 @@
+from smsframework.exc import ProviderError, RequestError
+
+class AfricasTalkingProviderError(ProviderError):
+    
+    def __init__(self, message):
+        super().__init__(message)
```
Comment (panoskar86) [inline smsframework_africastalking/error.py]: you need to press enter and add one more empty line(PEP-8) :)

File: smsframework_africastalking/provider.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+            sent_message = api_response['SMSMessageData']['Recipients'][0]
+
+            if sent_message['status'] != 'Success':
+                raise Exception
+        except:
+            raise AfricasTalkingProviderError('API_ERROR')
+
+        message.msgid = sent_message['messageId']
+        return message
```
Comment (panoskar86) [inline smsframework_africastalking/provider.py]: PEP-8 :)

File: smsframework_africastalking/provider.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+                message.dst,
+                message.provider_params['target_country']
+            )
+
+            formatted_number = phonenumbers.format_number(
+                phone_number,
+                phonenumbers.PhoneNumberFormat.E164
+            )
+        except:
+            raise AfricasTalkingProviderError('PHONE_NUMBER_PARSE_ERROR')
```
Comment (panoskar86) [inline smsframework_africastalking/provider.py:59]: do we need to log anything?

**Review by sramonsanz — APPROVED** [review]
Looks great apart from few PEP-8 stuff under `africastalking_test.py`.

File: tests/africastalking_test.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        )
+        message_back = self.gw.send(message_out)
+        self.assertEqual(message_back.msgid, '001')
+
+    @_mock_response(400, {'status': 'Failed'})
+    def test_send_failure(self):
+        message_out = OutgoingMessage('+254789789789', 'Hello Kenya',
+            provider='africas_talking'
+        )
+        self.assertRaises(AfricasTalkingProviderError, self.gw.send, message_out)
```
Comment (panoskar86) [inline tests/africastalking_test.py]: PEP-8

**Review by panoskar86 — CHANGES_REQUESTED** [review]
looks good. some really small changes for PEP-8 and I asked some questions

File: smsframework_africastalking/provider.py (hunk 4 above)
Comment (JoshBBOXX) [reply smsframework_africastalking/provider.py:59]: Added some generic logging

File: requirements.txt (hunk 1 above)
Comment (JoshBBOXX) [reply requirements.txt]: Changed to lite lib

File: smsframework_africastalking/error.py (hunk 2 above)
Comment (JoshBBOXX) [reply smsframework_africastalking/error.py]: done

File: smsframework_africastalking/provider.py (hunk 3 above)
Comment (JoshBBOXX) [reply smsframework_africastalking/provider.py]: done

File: tests/africastalking_test.py (hunk 5 above)
Comment (JoshBBOXX) [reply tests/africastalking_test.py]: done

**Review by panoskar86 — APPROVED** [review]
:+1:

**Pull request merged**